### PR TITLE
More resource documentation improvements

### PR DESCRIPTION
**Pull request by tas50**
Mostly just formatting differs from the docs side except for chef_handler which got some significant changes to make it *potentially* work IRL.

### CODE REVIEW

File: lib/chef/resource/chef_handler.rb [hunk 1]
```diff
@@ -25,8 +25,152 @@ class ChefHandler < Chef::Resource
...
+        source 'chef/handler/json_file'
+        arguments path: '/var/chef/reports'
+        type exception: true
+        action :enable
+      end
+      ```
+
+      **Cookbook Versions (a custom handler)**
+
+      juliandunn created a custom report handler that logs all of the cookbooks and cookbook versions that were used during a Chef Infra Client run, and then reports after the run is complete. This handler requires the chef_handler resource (which is available from the chef_handler cookbook).
```
Comment (phiggins) [inline lib/chef/resource/chef_handler.rb:67]: Should this link to julian's github profile or something?

Comment (phiggins) [inline lib/chef/resource/chef_handler.rb:67]:
```suggestion
      juliandunn created a custom report handler that logs all of the cookbooks and cookbook versions that were used during a #{ChefUtils::Dist::Infra::PRODUCT} run, and then reports after the run is complete. This handler requires the `chef_handler` resource (which is available from the `chef_handler` cookbook).
```

File: lib/chef/resource/chef_handler.rb [hunk 2]
```diff
@@ -25,8 +25,152 @@ class ChefHandler < Chef::Resource
...
+
+      cookbook_versions.rb:
+
+      The following custom handler defines how cookbooks and cookbook versions that are used during a Chef Infra Client run will be compiled into a report using the `Chef::Log` class in Chef Infra Client:
```
Comment (phiggins) [inline lib/chef/resource/chef_handler.rb:71]:
```suggestion
      The following custom handler defines how cookbooks and cookbook versions that are used during a #{ChefUtils::Dist::Infra::PRODUCT} run will be compiled into a report using the `Chef::Log` class in #{ChefUtils::Dist::Infra::PRODUCT}:
```

File: lib/chef/resource/chef_handler.rb [hunk 3]
```diff
@@ -25,8 +25,152 @@ class ChefHandler < Chef::Resource
...
+
+      ```ruby
+      require 'chef/log'
+
+      module Chef
+        class CookbookVersionsHandler < Chef::Handler
+          def report
+            cookbooks = run_context.cookbook_collection
+            Chef::Log.info('Cookbooks and versions run: #{cookbooks.keys.map {|x| cookbooks[x].name.to_s + ' ' + cookbooks[x].version} }')
```
Comment (phiggins) [inline lib/chef/resource/chef_handler.rb:80]:
```suggestion
            Chef::Log.info('Cookbooks and versions run: #{cookbooks.map {|x| x.name.to_s + ' ' + x.version }}')
```

File: lib/chef/resource/chef_handler.rb [hunk 4]
```diff
@@ -25,8 +25,152 @@ class ChefHandler < Chef::Resource
...
+          end
+        end
+      end
+      ```
+
+      default.rb:
+
+      The following recipe is added to the run-list for every node on which a list of cookbooks and versions will be generated as report output after every Chef Infra Client run.
```
Comment (phiggins) [inline lib/chef/resource/chef_handler.rb:88]:
```suggestion
      The following recipe is added to the run-list for every node on which a list of cookbooks and versions will be generated as report output after every #{ChefUtils::Dist::Infra::PRODUCT} run.
```

File: lib/chef/resource/chef_handler.rb [hunk 5]
```diff
@@ -25,8 +25,152 @@ class ChefHandler < Chef::Resource
...
+
+      [2013-11-26T03:11:06+00:00] INFO: Chef Infra Client Run complete in 0.300029878 seconds
+      [2013-11-26T03:11:06+00:00] INFO: Running report handlers
+      [2013-11-26T03:11:06+00:00] INFO: Cookbooks and versions run: ["cookbook_versions_handler 1.0.0"]
+      [2013-11-26T03:11:06+00:00] INFO: Report handlers complete
+      ```
+
+      **JsonFile Handler**
+
+      The json_file handler is available from the chef_handler cookbook and can be used with exceptions and reports. It serializes run status data to a JSON file. This handler may be enabled in one of the following ways.
```
Comment (phiggins) [inline lib/chef/resource/chef_handler.rb:113]:
```suggestion
      The json_file handler is available from the `chef_handler` cookbook and can be used with exceptions and reports. It serializes run status data to a JSON file. This handler may be enabled in one of the following ways.
```

File: lib/chef/resource/chef_handler.rb [hunk 6]
```diff
@@ -25,8 +25,152 @@ class ChefHandler < Chef::Resource
...
+
+      By adding the following lines of Ruby code to either the client.rb file or the solo.rb file, depending on how Chef Infra Client is being run:
```
Comment (phiggins) [inline lib/chef/resource/chef_handler.rb:115]:
```suggestion
      By adding the following lines of Ruby code to either the client.rb file or the solo.rb file, depending on how #{ChefUtils::Dist::Infra::PRODUCT} is being run:
```

File: lib/chef/resource/chef_handler.rb [hunk 7]
```diff
@@ -25,8 +25,152 @@ class ChefHandler < Chef::Resource
...
+
+      ```ruby
+      require 'chef/handler/json_file'
+      report_handlers << Chef::Handler::JsonFile.new(path: '/var/chef/reports')
+      exception_handlers << Chef::Handler::JsonFile.new(path: '/var/chef/reports')
+      ```
+
+      By using the chef_handler resource in a recipe, similar to the following:
```
Comment (phiggins) [inline lib/chef/resource/chef_handler.rb:123]:
```suggestion
      By using the `chef_handler` resource in a recipe, similar to the following:
```

File: lib/chef/resource/chef_handler.rb [hunk 8]
```diff
@@ -25,8 +25,152 @@ class ChefHandler < Chef::Resource
...
+      chef_handler 'Chef::Handler::JsonFile' do
+        source 'chef/handler/json_file'
+        arguments path: '/var/chef/reports'
+        action :enable
+      end
+      ```
+
+      **ErrorReport Handler**
+
+      The error_report handler is built into Chef Infra Client and can be used for both exceptions and reports. It serializes error report data to a JSON file. This handler may be enabled in one of the following ways.
```
Comment (phiggins) [inline lib/chef/resource/chef_handler.rb:155]:
```suggestion
      The error_report handler is built into #{ChefUtils::Dist::Infra::PRODUCT} and can be used for both exceptions and reports. It serializes error report data to a JSON file. This handler may be enabled in one of the following ways.
```

File: lib/chef/resource/chef_handler.rb [hunk 9]
```diff
@@ -25,8 +25,152 @@ class ChefHandler < Chef::Resource
...
+
+      By adding the following lines of Ruby code to either the client.rb file or the solo.rb file, depending on how Chef Infra Client is being run:
```
Comment (phiggins) [inline lib/chef/resource/chef_handler.rb:157]:
```suggestion
      By adding the following lines of Ruby code to either the `client.rb` file or the `solo.rb` file, depending on how #{ChefUtils::Dist::Infra::PRODUCT} is being run:
```

File: lib/chef/resource/chef_handler.rb [hunk 10]
```diff
@@ -25,8 +25,152 @@ class ChefHandler < Chef::Resource
...
+
+      ```ruby
+      require 'chef/handler/error_report'
+      report_handlers << Chef::Handler::ErrorReport.new
+      exception_handlers << Chef::Handler::ErrorReport.new
+      ```
+
+      By using the chef_handler resource in a recipe, similar to the following:
```
Comment (phiggins) [inline lib/chef/resource/chef_handler.rb:165]:
```suggestion
      By using the `chef_handler` resource in a recipe, similar to the following:
```

File: lib/chef/resource/ohai_hint.rb [hunk 11]
```diff
@@ -26,6 +26,39 @@ class OhaiHint < Chef::Resource
 
       description "Use the **ohai_hint** resource to aid in configuration detection by passing hint data to Ohai."
       introduced "14.0"
+      examples <<~DOC
+      **Create a hint file**
+
+      ```ruby
+      ohai_hint 'example' do
+        content Hash[:a, 'test_content']
```
Comment (phiggins) [inline lib/chef/resource/ohai_hint.rb:34]:
```suggestion
        content a: 'test_content'
```

File: lib/chef/resource/ohai_hint.rb [hunk 12]
```diff
@@ -26,6 +26,39 @@ class OhaiHint < Chef::Resource
...
+      ```ruby
+      ohai_hint 'example' do
+        compile_time false
+      end
+      ```
+
+      **Delete a hint file**
+
+      ```ruby
+      ohai-hint 'example' do
```
Comment (phiggins) [inline lib/chef/resource/ohai_hint.rb:57]:
```suggestion
      ohai_hint 'example' do
```

File: lib/chef/resource/chef_handler.rb [hunk 13]
```diff
@@ -25,8 +25,152 @@ class ChefHandler < Chef::Resource
 
       provides(:chef_handler) { true }
 
-      description "Use the **chef_handler** resource to install or uninstall reporting/exception handlers."
+      description "Use the **chef_handler** resource to enable handlers during a #{ChefUtils::Dist::Infra::PRODUCT} run. The resource allows arguments to be passed to #{ChefUtils::Dist::Infra::PRODUCT}, which then applies the conditions defined by the custom handler to the node attribute data collected during a #{ChefUtils::Dist::Infra::PRODUCT} run, and then processes the handler based on that data.\nThe **chef_handler** resource is typically defined early in a node's run-list (often being the first item). This ensures that all of the handlers will be available for the entire #{ChefUtils::Dist::Infra::PRODUCT} run."
       introduced "14.0"
+      examples <<~'DOC'
+      **Enable the 'MyHandler' handler**
+
+      The following example shows how to enable a fictional 'MyHandler' handler which is located on disk at `/etc/chef/my_handler.rb`. The handler will be configured to run with Chef Infra Client and will be passed values to the handler's initializer method:
```
Comment (phiggins) [inline lib/chef/resource/chef_handler.rb:33]:
```suggestion
      The following example shows how to enable a fictional 'MyHandler' handler which is located on disk at `/etc/chef/my_handler.rb`. The handler will be configured to run with #{ChefUtils::Dist::Infra::PRODUCT} and will be passed values to the handler's initializer method:
```

File: lib/chef/resource/bff_package.rb [hunk 14]
```diff
@@ -27,6 +27,27 @@ class BffPackage < Chef::Resource::Package
 
       description "Use the **bff_package** resource to manage packages for the AIX platform using the installp utility. When a package is installed from a local file, it must be added to the node using the **remote_file** or **cookbook_file** resources."
       introduced "12.0"
+      examples <<~DOC
+      The **bff_package** resource is the default package provider on the AIX platform. The base **package** resource may be used, and then when the platform is AIX, Chef Infra Client will identify the correct package provider. The following examples show how to install part of the IBM XL C/C++ compiler.
```
Comment (phiggins) [inline lib/chef/resource/bff_package.rb:31]:
```suggestion
      The **bff_package** resource is the default package provider on the AIX platform. The base **package** resource may be used, and then when the platform is AIX, #{ChefUtils::Dist::Infra::PRODUCT} will identify the correct package provider. The following examples show how to install part of the IBM XL C/C++ compiler.
```

**Review by phiggins — APPROVED** [review]

File: lib/chef/resource/chef_handler.rb (hunk 1 above)
Comment (tas50) [reply lib/chef/resource/chef_handler.rb:67]: I had issues using the dist values within this heredoc so I instead chose to treat this as a plain text block without any interpolation.

Comment (tas50) [reply lib/chef/resource/chef_handler.rb]: Added

Comment (phiggins) [reply lib/chef/resource/chef_handler.rb]: Weird, probably easy enough to search-and-replace if we ever wanted to.

**Pull request merged**